src/modules/connection.js

```diff
@@ -27,6 +29,7 @@ class Connection {
 
   async _connect() {
     try {
+      await this.hooks.trigger(this, ConnectionHooks.beforeConnectEvent, { config: this._config });
```

Comment:
should we do `Object.freeze` on config? as i don't think we want hook to change configs?

Reply:
I froze the config, modified the tests that mutated it, and moved the logger from it to it's own module, so not everything that wants to log (like the hooks) depend on the root configuration.